src/_pytest/cacheprovider.py

```diff
@@ -178,20 +183,13 @@ def set(self, key: str, value: object) -> None:
         """
         path = self._getvaluepath(key)
         try:
-            if path.parent.is_dir():
-                cache_dir_exists_already = True
-            else:
-                cache_dir_exists_already = self._cachedir.exists()
-                path.parent.mkdir(exist_ok=True, parents=True)
+            self._mkdir(path.parent)
         except OSError as exc:
             self.warn(
                 f"could not create cache path {path}: {exc}",
                 _ispytest=True,
             )
             return
-        if not cache_dir_exists_already:
-            self._ensure_supporting_files()
-        data = json.dumps(value, ensure_ascii=False, indent=2)
```

Comment:
This code is not *exactly* the same as before, previously if we failed to encode the data to JSON we would not write a file at all, but as is we will have an empty file instead.

Given you mentioned this was for a small performance improvement (which I suspect is not there anyway, would have to benchmark it), let's revert to the previous behavior to ensure we keep the same behavior. 👍

Reply:
Done.